### ISSUE TITLE
Deserialize in .NET

### ISSUE THREAD

**Issue by Marbry** [cubiclesoft/email_sms_mms_gateways#7]
This was a little more effort than I anticipated, but here is the code in case anyone is interested.  There may be a cleaner way to accomplish this without explicitly looping through each list, if so I'd be curious to see how.

I used Json.Net, credit to this site http://csharphelper.com/blog/2013/05/how-to-extract-only-some-of-the-information-from-a-json-file-in-c/  for some of the approach.

```c#
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

public Dictionary<string, CountryInfo> GetCarriers()
{
    // Get the data file.
    string jsonPath = @"your path here";
    string serialization = "";

    Dictionary<string, CountryInfo> Country = new Dictionary<string, CountryInfo>();
    Dictionary<string, Dictionary<string, CarrierInfo>> SMS_Carriers = new Dictionary<string,Dictionary<string,CarrierInfo>>();
    Dictionary<string, Dictionary<string, CarrierInfo>> MMS_Carriers = new Dictionary<string, Dictionary<string, CarrierInfo>>();

    StreamReader sr = File.OpenText(jsonPath);
    serialization = sr.ReadToEnd();

    Dictionary<string, object> dict = JsonConvert.DeserializeObject<Dictionary<string, object>>(serialization);

    // Get the countries.
    Dictionary<string, string> countries = JsonConvert.DeserializeObject<Dictionary<string, string>>(dict["countries"].ToString());

    foreach (KeyValuePair<string, string> pair in countries)
    {
        CountryInfo country_info = new CountryInfo(pair.Key, (string)pair.Value, new List<CarrierInfo>(), new List<CarrierInfo>());
        Country.Add(country_info.CountryAbbreviation, country_info);
    }

    // Get the SMS carriers.
    Dictionary<string, Dictionary<string, string[]>> jsonCarriers = JsonConvert.DeserializeObject<Dictionary<string, Dictionary<string, string[]>>>(dict["sms_carriers"].ToString());

    foreach (string countryAbbr in jsonCarriers.Keys)
    {
        Dictionary<string, CarrierInfo> smscarriers = new Dictionary<string, CarrierInfo>();

        foreach (string carrierAbbr in jsonCarriers[countryAbbr].Keys)
        {
            CarrierInfo carrier = new CarrierInfo(jsonCarriers[countryAbbr][carrierAbbr][0], carrierAbbr, jsonCarriers[countryAbbr][carrierAbbr].Skip(1).ToList());
            smscarriers.Add(carrier.CarrierAbbreviation, carrier);
        }

        SMS_Carriers.Add(countryAbbr, smscarriers);
    }

    // Get the MMMS carriers
    jsonCarriers = JsonConvert.DeserializeObject<Dictionary<string, Dictionary<string, string[]>>>(dict["mms_carriers"].ToString());

    foreach (string countryAbbr in jsonCarriers.Keys)
    {
        Dictionary<string, CarrierInfo> MMScarriers = new Dictionary<string, CarrierInfo>();

        foreach (string carrierAbbr in jsonCarriers[countryAbbr].Keys)
        {
            CarrierInfo carrier = new CarrierInfo(jsonCarriers[countryAbbr][carrierAbbr][0], carrierAbbr, jsonCarriers[countryAbbr][carrierAbbr].Skip(1).ToList());
            MMScarriers.Add(carrier.CarrierAbbreviation, carrier);
        }

        MMS_Carriers.Add(countryAbbr, MMScarriers);
    }

    // Populate return object
    foreach (string countryAbbr in Country.Keys)
    {
        Country[countryAbbr].SMSCarriers.AddRange(SMS_Carriers[countryAbbr].Values.ToList());
        if (MMS_Carriers.ContainsKey(countryAbbr))
            Country[countryAbbr].MMSCarriers.AddRange(MMS_Carriers[countryAbbr].Values.ToList());
    }

    return Country;
}


public class CarrierInfo
{
    public string CarrierAbbreviation { get; set; }
    public string CarrierName { get; set; }
    public List<string> Emails { get; set; }

    public CarrierInfo(string carrierName, string carrierAbbreviation, List<string> emails)
    {
        CarrierName = carrierName;
        CarrierAbbreviation = carrierAbbreviation;
        Emails = emails;
    }

    public override string ToString()
    {
        return CarrierName;
    }
}


public class CountryInfo
{
    public string CountryAbbreviation {get; set;}
    public string CountryName { get; set; }
    public List<CarrierInfo> SMSCarriers { get; set; }
    public List<CarrierInfo> MMSCarriers { get; set; }

    public CountryInfo(string countryAbbreviation, string countryName, List<CarrierInfo> smsCarriers, List<CarrierInfo> mmsCarriers)
    {
        CountryAbbreviation = countryAbbreviation;
        CountryName = countryName;
        SMSCarriers = smsCarriers;
        MMSCarriers = mmsCarriers;
    }

    public override string ToString()
    {
        return CountryName;
    }
}
```

**Comment by cubiclesoft** [comment]
Json.NET is the defacto standard JSON parser for .NET languages.  Your code looks okay to me.

I'm not sure what I want to do with this.  It's not really maintainable code in its current form.  If you make a GitHub repo for it, I could link to it from this repo so those looking for relevant code snippets can use the above.  I maintain a C++ snippets library:

https://github.com/cubiclesoft/cross-platform-cpp

Basically, you could do something similar where you dump C# snippets like the one in this issue.  Snippet libraries are less formal than a full project - just a collection of random functions/classes that don't need a whole repo to themselves.  I like to isolate repos so I don't have random junk floating around.
